internal/lru/lru.go

```diff
@@ -34,64 +38,120 @@ type Cache struct {
+	mu     sync.RWMutex
 }
 
 type entry struct {
-	key   string
-	value interface{}
+	key       string
+	value     interface{}
+	timestamp time.Time
 }
 
 // New creates a new Cache.
 // If maxEntries is zero, the cache has no limit and it's assumed
 // that eviction is done by the caller.
-func New(maxEntries int) *Cache {
+// If window is zero the "windowed updates" of the LRU list will be disabled.
+// For further information on this technique see: https://www.openmymind.net/High-Concurrency-LRU-Caching/
+func New(maxEntries int, window time.Duration) *Cache {
 	return &Cache{
 		MaxEntries: maxEntries,
 		ll:         list.New(),
 		cache:      make(map[string]*list.Element),
+		window:     window,
 	}
 }
 
 // Add adds a value to the cache.
-func (c *Cache) Add(key string, value interface{}) {
-	if c.cache == nil {
-		c.cache = make(map[string]*list.Element)
-		c.ll = list.New()
-	}
-	if ee, ok := c.cache[key]; ok {
-		c.ll.MoveToFront(ee)
-		ee.Value.(*entry).value = value
+func (c *Cache) Add(key string, val interface{}) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if ele, ok := c.cache[key]; ok {
+		c.ll.MoveToFront(ele)
+		ele.Value.(*entry).value = val
 		return
 	}
-	ele := c.ll.PushFront(&entry{key, value})
+	c.addLocked(key, val)
+}
+
+func (c *Cache) addLocked(key string, val interface{}) {
+	timestamp := time.Now()
+	ent := &entry{key, val, timestamp}
+	ele := c.ll.PushFront(ent)
 	c.cache[key] = ele
 	if c.MaxEntries != 0 && c.ll.Len() > c.MaxEntries {
-		c.RemoveOldest()
+		c.removeOldestLocked()
 	}
 }
 
 // Get looks up a key's value from the cache.
 func (c *Cache) Get(key string) (value interface{}, ok bool) {
-	if c.cache == nil {
-		return
+	timestamp := time.Now()
+	c.mu.RLock()
+	if ele, ok := c.cache[key]; ok {
+		ent := ele.Value.(*entry)
+		if timestamp.After(ent.timestamp.Add(c.window)) {
+			c.mu.RUnlock()
+			c.mu.Lock()
+			// We dropped the lock so we need to perform the lookup again.
+			if ele, ok := c.cache[key]; ok {
+				ent := ele.Value.(*entry)
+				ent.timestamp = timestamp
+				c.ll.MoveToFront(ele)
+				c.mu.Unlock()
+			} else {
+				c.mu.Unlock()
+				return nil, false
+			}
+		} else {
+			c.mu.RUnlock()
+		}
+		return ent.value, true
 	}
-	if ele, hit := c.cache[key]; hit {
-		c.ll.MoveToFront(ele)
+	c.mu.RUnlock()
+	return
+}
+
+// GetOrInsert either retrieves the current entry for the key or inserts the
+// provided value for the key.
+// It returns true if the value was loaded and false if the provided value
+// was stored.
+func (c *Cache) GetOrInsert(key string, val interface{}) (interface{}, bool) {
+	// Optimistically try a Get first
+	value, ok := c.Get(key)
+	if ok {
+		return value, ok
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// We need to check again
+	if ele, ok := c.cache[key]; ok {
 		return ele.Value.(*entry).value, true
 	}
-	return
+
+	c.addLocked(key, val)
+	return val, false
+}
+
+// Len returns the number of items in the cache.
+func (c *Cache) Len() int {
+	c.mu.RLock()
```

Comment:
You could use `defer c.mu.RUnlock()` here.

Reply:
It felt so short and readable anyway. No need to push a defer on the stack.

Reply:
Sure.